File: src/api/daemon/analytics.ts
/**
 * Analytics API module — `POST /analytics/capture` (ADR-008 D20).
 *
 * Analytics API 模块 — `POST /analytics/capture`（ADR-008 D20）。
 *
 * The daemon is the single authoritative product-analytics sender. The GUI
 * webview routes its UI-interaction events through this endpoint instead of an
 * in-process sink so the two processes never double-count device-level signals
 * in PostHog.
 *
 * daemon 是 product analytics 的唯一权威发送方。GUI webview 的 UI 交互事件经
 * 此端点上报，而非进程内 sink，避免两进程在 PostHog 重复计数设备级信号。
 */

import { captureUiEvent as captureUiEventSdk } from '@/api/generated/sdk.gen'
import type { CaptureUiEventRequest } from '@/api/generated/types.gen'
import { daemonClient } from './client'

export type { CaptureUiEventRequest }

/**
 * Capture a GUI UI-interaction event via the daemon.
 *
 * Fire-and-forget on the wire: the resolved value only confirms the daemon
 * decoded the event and handed it to its sink, not that it reached PostHog.
 *
 * @param event Discriminated-union event payload (by `kind`).
 * @throws {DaemonApiError} On HTTP or session errors.
 */
export async function captureUiEvent(event: CaptureUiEventRequest): Promise<void> {
  await daemonClient.callSdk(() => captureUiEventSdk({ body: event, throwOnError: true }))
}

File: crates/uc-daemon-client/src/http/analytics.rs
use std::sync::Arc;

use anyhow::{anyhow, Context, Result};
use reqwest::{Method, RequestBuilder};

use crate::http::authorized_daemon_request_with_type;
use crate::DaemonConnectionState;
use uc_daemon_contract::api::dto::analytics::CaptureUiEventRequest;

/// Client for `POST /analytics/capture` (ADR-008 D20).
///
/// The daemon is the single authoritative product-analytics sender. UI-process
/// events that originate in the GUI's own Rust background tasks (the updater /
/// scheduler, which the webview never sees) are forwarded here so the daemon
/// dispatches them through its own gated sink + `EventContext`, instead of the
/// GUI holding a second in-process PostHog sink that would double-count
/// device-level signals.
///
/// Fire-and-forget on the wire: a successful call only confirms the daemon
/// decoded the event and handed it to its sink, not that it reached PostHog.
#[derive(Clone)]
pub struct DaemonAnalyticsClient {
    http: Arc<reqwest::Client>,
    connection_state: DaemonConnectionState,
    client_type: String,
}

impl DaemonAnalyticsClient {
    pub fn new(connection_state: DaemonConnectionState) -> Self {
        Self {
            http: Arc::new(reqwest::Client::new()),
            connection_state,
            client_type: "gui".to_string(),
        }
    }

    /// POST a single UI-interaction event to the daemon. Returns `Err` on a
    /// missing connection, transport failure, or non-2xx status — callers
    /// treat analytics as best-effort and should only `debug`-log the error.
    pub async fn capture(&self, event: CaptureUiEventRequest) -> Result<()> {
        let response = self
            .authorized_request(Method::POST, "/analytics/capture")
            .await?
            .json(&event)
            .send()
            .await
            .with_context(|| "failed to call daemon /analytics/capture")?;

        if response.status().is_success() {
            return Ok(());
        }
        let status = response.status();
        let body = response
            .text()
            .await
            .unwrap_or_else(|_| "<failed to read body>".to_string());
        Err(anyhow!(
            "daemon /analytics/capture failed with status {}: {}",
            status,
            body,
        ))
    }

    async fn authorized_request(&self, method: Method, path: &str) -> Result<RequestBuilder> {
        let connection = self
            .connection_state
            .get()
            .ok_or_else(|| anyhow!("daemon connection info is not available"))?;
        authorized_daemon_request_with_type(
            &self.http,
            &self.connection_state,
            method,
            path,
            connection.pid,
            &self.client_type,
        )
        .await
    }
}
